### ISSUE TITLE
wgengine/magicsock: fix disco cache bugs when a node key changes

### ISSUE THREAD

**Issue by bradfitz** [tailscale/tailscale#1391]
When a peer changes node keys, a few caches in Conn cache things based on discoKey:

```go
        nodeOfDisco map[tailcfg.DiscoKey]*tailcfg.Node
        discoOfNode map[tailcfg.NodeKey]tailcfg.DiscoKey
        discoOfAddr map[netaddr.IPPort]tailcfg.DiscoKey // validated non-DERP paths only                                                                                                 
        // endpointsOfDisco tracks the wireguard-go endpoints for peers                                                                                                                  
        // with recent activity.                                                                                                                                                         
        endpointOfDisco map[tailcfg.DiscoKey]*discoEndpoint // those with activity only                                                                                                  
        sharedDiscoKey  map[tailcfg.DiscoKey]*[32]byte      // nacl/box precomputed key                                                                                                  
```

But if a node changes its node key, we don't update those. We only clean up entries up when the disco key disappears:

```go
        // Clean c.endpointOfDisco for discovery keys that are no longer present.                                                                                                        
        for dk, de := range c.endpointOfDisco {
                if _, ok := c.nodeOfDisco[dk]; !ok {
                        de.stopAndReset()
                        delete(c.endpointOfDisco, dk)
                        delete(c.sharedDiscoKey, dk)
                }
        }
```

/cc @crawshaw @darshinimashar @danderson

**Comment by bradfitz** [comment]
In particular, looks like it's:

```go
        endpointOfDisco map[tailcfg.DiscoKey]*discoEndpoint // those with activity only                                                                                                  
```

We do try to update the discoEndpoint when its node changes:

```go
                if ep, ok := c.endpointOfDisco[n.DiscoKey]; ok {
                        ep.updateFromNode(n)
                }
```

But that `updateFromNode` doesn't handle its node key changing. In fact, discoEndpoint calls out that a node key can never change:

```go
// discoEndpoint is a wireguard/conn.Endpoint for new-style peers that                                                                                                                                                      
// advertise a DiscoKey and participate in active discovery.                                                                                                                                                                
type discoEndpoint struct {
        // atomically accessed; declared first for alignment reasons                                                                                                                                                        
        lastRecvUnixAtomic    int64
        numStopAndResetAtomic int64

        // These fields are initialized once and never modified.                                                                                                                                                            
        c                  *Conn
        publicKey          tailcfg.NodeKey  // peer public key (for WireGuard + DERP)                                                                                                                                       
        discoKey           tailcfg.DiscoKey // for discovery mesages                                                                                                                                                        
        discoShort         string           // ShortString of discoKey                                                                                                                                                      
        fakeWGAddr         netaddr.IPPort   // the UDP address we tell wireguard-go we're using                                                                                                                             
        wgEndpointHostPort string           // string from CreateEndpoint: "<hex-discovery-key>.disco.tailscale:12345"                                                                                                      
```

**Comment by bradfitz** [comment]
Example failing test. The `workaround` const demos what control could do, @crawshaw.

```go
func TestSetNetworkMapChangingNodeKey(t *testing.T) {
        conn := newNonLegacyTestConn(t)
        t.Cleanup(func() { conn.Close() })
        conn.SetPrivateKey(wgkey.Private{0: 1})

        discoKey := tailcfg.DiscoKey{31: 1}
        discoKey2 := tailcfg.DiscoKey{31: 2}
        nodeKey1 := tailcfg.NodeKey{0: 'N', 1: 'K', 2: '1'}
        nodeKey2 := tailcfg.NodeKey{0: 'N', 1: 'K', 2: '2'}

        conn.SetNetworkMap(&netmap.NetworkMap{
                Peers: []*tailcfg.Node{
                        {
                                Key:       nodeKey1,
                                DiscoKey:  discoKey,
                                Endpoints: []string{"192.168.1.2:345"},
                        },
                },
        })
        ep, err := conn.CreateEndpoint([32]byte(nodeKey1), "0000000000000000000000000000000000000000000000000000000000000001.disco.tailscale:12345")
        if err != nil {
                t.Fatal(err)
        }
        t.Logf("Got: %T %p", ep, ep)

        const workaround = false
        if workaround {
                conn.SetNetworkMap(&netmap.NetworkMap{
                        Peers: []*tailcfg.Node{
                                {
                                        Key:       nodeKey2,
                                        DiscoKey:  discoKey2,
                                        Endpoints: []string{"192.168.1.2:345"},
                                },
                        },
                })
        }
        conn.SetNetworkMap(&netmap.NetworkMap{
                Peers: []*tailcfg.Node{
                        {
                                Key:       nodeKey2,
                                DiscoKey:  discoKey,
                                Endpoints: []string{"192.168.1.2:345"},
                        },
                },
        })

        if workaround {
                ep2, err := conn.CreateEndpoint([32]byte(nodeKey2), "0000000000000000000000000000000000000000000000000000000000000001.disco.tailscale:12345")
                if err != nil {
                        t.Fatal(err)
                }
                t.Logf("Got: %T %p", ep2, ep2)
        }

        de := conn.endpointOfDisco[discoKey]
        if de == nil {
                t.Fatal("no discoEndpoint for discoKey")
        }
        if de.publicKey != nodeKey2 {
                t.Fatalf("discoEndpoint public key = %02x; want %02x", de.publicKey, nodeKey2)
        }
}
```

**Comment by bradfitz** [comment]
This fix will be in 1.4.5.

In the meantime, clients need to restart to get unstuck. (On Windows, this involves restarting the service, not just the GUI process.)
